### PR TITLE
Fixed the errors for '-Werror' flag

### PR DESCRIPTION
**Pull request by aprokop**
Mostly works, but disabling it for STK for now as it has issues. I also
have not tried with other adapters like Moab.
    
One needs to configure with `-Werror -Wall` but without `-Wextra
-Weffc++`.

Addresses #130.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -13,6 +13,8 @@ TRIBITS_PACKAGE_DECL(DataTransferKit)
 TRIBITS_ADD_DEBUG_OPTION()
 TRIBITS_ADD_SHOW_DEPRECATED_WARNINGS_OPTION()
 
+SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror")
+
```
Comment (Rombur) [inline CMakeLists.txt]: This will affect everybody, right? Including user? I would prefer if we would only use the flag for jenkins. Also does `-Werror` work for every compiler?

File: packages/Adapters/BasicGeometry/test/tstCylinderGeometry.cpp [hunk 2]
```diff
@@ -123,7 +123,7 @@ TEUCHOS_UNIT_TEST( CylinderGeometry, cylinder_test )
         double y_distance = 0.0;
         double centroid_distance = 0.0;
         double tol = 1.0e-6;
-        for ( int i = 0; i < num_rand; ++i )
+        for ( int k = 0; k < num_rand; ++k )
```
Comment (Rombur) [inline packages/Adapters/BasicGeometry/test/tstCylinderGeometry.cpp]: Any reason for this change?

File: packages/Adapters/ClassicDTK/src/DTK_VolumeSourceMap_def.hpp [hunk 3]
```diff
@@ -171,7 +171,7 @@ void VolumeSourceMap<Geometry, GlobalOrdinal, CoordinateField>::setup(
         Teuchos::ArrayRCP<GlobalOrdinal> source_gids =
             source_geometry_manager->gids();
         int local_num_source_geom = source_geometry_manager->localNumGeometry();
-        for ( std::size_t i = 0; i < local_num_source_geom; ++i )
+        for ( int i = 0; i < local_num_source_geom; ++i )
```
Comment (Rombur) [inline packages/Adapters/ClassicDTK/src/DTK_VolumeSourceMap_def.hpp]: Shouldn't you change `local_num_source_geom`? It doesn't make sense for this to be negative.

File: packages/Adapters/ClassicDTK/test/geometry/tstBox.cpp [hunk 4]
```diff
@@ -112,7 +112,7 @@ TEUCHOS_UNIT_TEST( Box, box_test )
         Teuchos::Array<double> point( 3 );
         int num_rand = 100;
         double tol = 1.0e-6;
-        for ( int i = 0; i < num_rand; ++i )
+        for ( int k = 0; k < num_rand; ++k )
```
Comment (Rombur) [inline packages/Adapters/ClassicDTK/test/geometry/tstBox.cpp]: Why?

**Comment by aprokop** [comment]
I'm not sure why the tests are **Not Run**. It seems the code's been built fine.

File: packages/Utils/src/Nanoflann/DTK_nanoflann.hpp [hunk 5]
```diff
@@ -618,13 +618,13 @@ class PooledAllocator
             wastedMemory += remaining;
 
             /* Allocate new storage. */
-            const size_t blocksize =
+            const size_t blk_size =
```
Comment (Rombur) [inline packages/Utils/src/Nanoflann/DTK_nanoflann.hpp]: Why the change?

File: CMakeLists.txt (hunk 1 above)
Comment (aprokop) [inline CMakeLists.txt]: Well, if I understand CMake correctly, it will only append `-Werror` while building DTK code, and not otherwise. I may be wrong about it.

File: packages/Adapters/BasicGeometry/test/tstCylinderGeometry.cpp (hunk 2 above)
Comment (aprokop) [inline packages/Adapters/BasicGeometry/test/tstCylinderGeometry.cpp]: There is an outer loop with `i` index, this internal loop hid `i`

File: packages/Adapters/ClassicDTK/test/geometry/tstBox.cpp (hunk 4 above)
Comment (aprokop) [inline packages/Adapters/ClassicDTK/test/geometry/tstBox.cpp]: There is an outer loop with `i` index, this internal loop hid `i`

File: packages/Utils/src/Nanoflann/DTK_nanoflann.hpp (hunk 5 above)
Comment (aprokop) [inline packages/Utils/src/Nanoflann/DTK_nanoflann.hpp]: `blocksize` hid another one outside

File: packages/Adapters/ClassicDTK/src/DTK_VolumeSourceMap_def.hpp (hunk 3 above)
Comment (aprokop) [inline packages/Adapters/ClassicDTK/src/DTK_VolumeSourceMap_def.hpp]: Well, it was one or the other. I chose a simpler one.

File: CMakeLists.txt (hunk 1 above)
Comment (aprokop) [inline CMakeLists.txt]: No, don't get me wrong, I agree with you in that we should not hard-code this. I thought you had a question of how it works. I'm pretty sure it won't work on every compiler, as some of them have pretty exotic flags.

Comment (Rombur) [inline CMakeLists.txt]: I think that neams wants to use `xlc` that's why I ask. If we decide that we don't support others compilers (I don't DTK has been built with any compilers but gcc, clang, and intel), then hard-coding the value is fine.

Comment (dalg24) [inline CMakeLists.txt]: No.  Configure with `-D CMAKE_CXX_FLAGS=-Werror` if you like but do not add it here.  We certainly do not want to pass the option for all DTK builds.
That said, I am fine with adding it to `scripts/docker_cmake`.

File: packages/Adapters/STKMesh/CMakeLists.txt [hunk 6]
```diff
@@ -15,6 +15,11 @@ TRIBITS_SUBPACKAGE(STKMeshAdapters)
 ## Add library, test, and examples.
 ##---------------------------------------------------------------------------##
 
+# Remove unwanted flag
+STRING(REPLACE "-Werror" "" CUSTOM_FLAGS ${CMAKE_CXX_FLAGS})
+SET(CMAKE_CXX_FLAGS "${CUSTOM_FLAGS}")
```
Comment (dalg24) [inline packages/Adapters/STKMesh/CMakeLists.txt]: Does that mean that the STKMesh adapters won't compile with that option?

Comment (aprokop) [inline packages/Adapters/STKMesh/CMakeLists.txt]: Yes, there are issues there.

File: CMakeLists.txt (hunk 1 above)
Comment (aprokop) [inline CMakeLists.txt]: I added [WIP] because of that. I wanted to run all the CI test first to make sure it actually works. This is the only way I could do that.

**Comment by dalg24** [comment]
@aprokop Have you looked at the [TriBITS documentation on compiler options](https://tribits.org/doc/TribitsBuildReference.html#selecting-compiler-and-linker-options)?
It looks like `-D DataTransferKit_CXX_FLAGS="-Werror"` would set the `CMAKE_CXX_FLAGS` variable for our package but that does not address the STKMesh issue.

**Comment by aprokop** [comment]
@dalg24 wrote:

> It looks like -D DataTransferKit_CXX_FLAGS="-Werror" would set the CMAKE_CXX_FLAGS variable for our package but that does not address the STKMesh issue.

I just checked and indeed, the `DataTransferKit_CXX_FLAGS` seems to simply append extra flags to `CMAKE_CXX_FLAGS`. I got lost in the Tribits document, though, and cannot find the place explicitly stating it.

**Comment by aprokop** [comment]
Reopened PR to restart CI.

**Comment by dalg24** [comment]
>  I got lost in the Tribits document, though, and cannot find the place explicitly stating it.

Look [here](https://tribits.org/doc/TribitsBuildReference.html#adding-arbitrary-compiler-flags-but-keeping-default-build-type-flags).

**Comment by aprokop** [comment]
Running in a docker container, I think I see what's going on. In fact, this PR even fails to build (that CI does not seem to show) due to the presence of `-Wextra` and `-Weffc++` flags. Without those in `docker_cmake`, it builds and passes fine if additional `-Wno-unknown-pragmas` is added.

The pragmas thing is because omp pragmas in some places are not recognized. Not sure what's going on there.

**Comment by aprokop** [comment]
Rebased to fix conflicts, and want to see the results of full CI log.

**Comment by aprokop** [comment]
That's weird, all the examples now ran without any changes...

OK, so we should now actually discuss what form this should be merged in. Obviously, the `SET(CXX_FLAGS "-Werror")` in `DTK/CMakeLists.txt` is going to be gone. 

We may need a second build (nightly?) to run a different build (other than CI) to test for warnings (with removed `-Wextra` and `-Weffc++`). In general, we should discuss if we want to have builds other than CI, and how to manage them.

**Comment by aprokop** [comment]
@dalg24 Please review.

File: packages/Adapters/ClassicDTK/src/DTK_VolumeSourceMap_def.hpp [hunk 7]
```diff
@@ -401,7 +401,8 @@ void VolumeSourceMap<Geometry, GlobalOrdinal, CoordinateField>::apply(
     {
         DTK_REQUIRE(
             target_field_view.size() ==
-            Teuchos::as<GlobalOrdinal>( d_target_map->getNodeNumElements() ) *
+            Teuchos::as<typename decltype( target_field_view )::size_type>(
+                d_target_map->getNodeNumElements() ) *
```
Comment (dalg24) [inline packages/Adapters/ClassicDTK/src/DTK_VolumeSourceMap_def.hpp]: That's the hard way here too...

File: packages/Adapters/STKMesh/CMakeLists.txt [hunk 8]
```diff
@@ -15,6 +15,11 @@ TRIBITS_SUBPACKAGE(STKMeshAdapters)
 ## Add library, test, and examples.
 ##---------------------------------------------------------------------------##
 
+# STK adapters do not work with "-Werror -Wall"
+STRING(REPLACE "-Werror" "" CUSTOM_FLAGS ${CMAKE_CXX_FLAGS})
+SET(CMAKE_CXX_FLAGS "${CUSTOM_FLAGS}")
+
+
```
Comment (dalg24) [inline packages/Adapters/STKMesh/CMakeLists.txt]: Not sure whether we should keep this.

Comment (aprokop) [inline packages/Adapters/STKMesh/CMakeLists.txt]: A different approach would be to disable STK mesh adapters when running with `-Werror`.

**Comment by dalg24** [comment]
@aprokop What happened between build#190 and build#191 [on Jenkins](https://jupyterdocker.ornl.gov:8080/job/DataTransferKit-experimental/)?
Did you revert your changes to the project `CMakeLists.txt` (or `./scripts/docker_cmake` I can't remember where you were adding `"-Werror"` to `CMAKE_CXX_FLAGS`).  It seems like you did not have `-Wextra` and `-Weffc++` in build#190.

**Comment by aprokop** [comment]
@dalg24 Nothing happened. The only thing I did was to rebase on top of the current master which had your jenkins changes merged (maybe something else too, not sure).

**Comment by dalg24** [comment]
I'd say go ahead and edit the [`CMAKE_CXX_FLAGS` variable in `scripts/docker_cmake`](https://github.com/ORNL-CEES/DataTransferKit/blob/master/scripts/docker_cmake#L7).  I'm good with all the changes but the one to `packages/Adapters/STKMesh/CMakeLists.txt`.  I'd keep it on a separate branch for whenever we implement a build with `-Werror`.

**Comment by aprokop** [comment]
What changes do you propose for `docker_cmake`?
I'm fine with removing things from STKMesh.

**Comment by aprokop** [comment]
I'm not going to do changes to `docker_cmake` as the build does not work without disabling Werror in STK adapter.

File: packages/Interface/src/OperatorVector/DTK_IntegrationPointSet.hpp [hunk 9]
```diff
@@ -80,6 +80,9 @@ class IntegrationPointSetIterator : public EntityIterator
     // Pre-increment operator.
     EntityIterator &operator++() override;
 
+    // Post-increment operator.
+    using EntityIterator::operator++;
```
Comment (sslattery) [inline packages/Interface/src/OperatorVector/DTK_IntegrationPointSet.hpp]: Isn't this functionality automatically inherited from the base class?

File: packages/Interface/test/tstEntityIterator.cpp [hunk 10]
```diff
@@ -193,6 +193,9 @@ class VectorIterator : public DataTransferKit::EntityIterator
         return *this;
     }
 
+    // Post-increment operator.
+    using DataTransferKit::EntityIterator::operator++;
```
Comment (sslattery) [inline packages/Interface/test/tstEntityIterator.cpp]: Same here on the inheritance of this function. What error did this result in?

File: packages/Interface/src/OperatorVector/DTK_IntegrationPointSet.hpp (hunk 9 above)
Comment (aprokop) [inline packages/Interface/src/OperatorVector/DTK_IntegrationPointSet.hpp]: It does not by default. If you have overloaded operator, the override version in inherited class hides all others by default. [1](javascript override overload function) [2](http://stackoverflow.com/questions/1628768/why-does-an-overridden-function-in-the-derived-class-hide-other-overloads-of-the)

File: packages/Interface/test/tstEntityIterator.cpp (hunk 10 above)
Comment (aprokop) [inline packages/Interface/test/tstEntityIterator.cpp]: The error was from the warning about hidden function, I believe.

**Comment by aprokop** [comment]
@sslattery Do you know how to fix
```bash
../../DataTransferKit/packages/Adapters/Libmesh/src/DTK_LibmeshEntitySet.cpp: In member function ‘virtual Teuchos::RCP<const Teuchos::Comm<int> > DataTransferKit::LibmeshEntitySet::communicator() const’:
../../DataTransferKit/packages/Adapters/Libmesh/src/DTK_LibmeshEntitySet.cpp:65:62: error: invalid conversion from ‘libMesh::Parallel::communicator {aka int}’ to ‘MPI_Comm {aka ompi_communicator_t*}’ [-fpermissiv
e]
         new Teuchos::MpiComm<int>( d_libmesh_mesh->comm().get() ) );
                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~^~
```

**Comment by aprokop** [comment]
@sslattery Nevermind, Spack installs `libmesh` without MPI by default. Have llnl/spack#2584 in progress.

**Comment by aprokop** [comment]
Updated PR with changes to Moab and libmesh adapters. It *may* work.

**Comment by aprokop** [comment]
Need trilinos/Trilinos#927 to finish.

**Comment by aprokop** [comment]
`-Werror` for DTK passes! 🏁 

Please do the (final?) review.
